certbot-dns-bookmyname/certbot_dns_bookmyname/dns_bookmyname.py

```diff
@@ -0,0 +1,330 @@
+
+        NOTE: this returns the raw zone content, with lots of spaces
+        inserted in order to align fields if using a fixed width font.
+
+        This method will log in if necessary
+        """
+        self.log_in()
+        try:
+            zone_page = self.session.get(
+                "https://www.bookmyname.com/manager.cgi?cmd=gdp&did={0}&mode=1"
+                .format(domain_id)
+            )
+            zone_html = BeautifulSoup(zone_page.text, 'html.parser')
+            zone_textarea = zone_html.select('textarea[name="gdp_zonefile"]')[0]
+            return zone_textarea.text.strip()
+        except:
+            raise ZoneError
+
+    def set_domain_zone(self, domain_id, zone):
+        """
+        Set the zone for a given domain name
+
+        NOTES: Extra whitespace within lines has no effect, but OTOH,
+        the zone cannot be entirely empty, so if it is, we replace it
+        with a single \n.
+        Also, TXT records (and possibly others) require double quotes.
+
+        This method will log in if necessary
+        """
+        if zone == '':
+            zone = '\n'
+        self.log_in()
+        try:
+            self.session.post(
+                "https://www.bookmyname.com/manager.cgi?cmd=gdp&mode=1",
+                {
+                    'did': domain_id,
+                    'gdp_zonefile': zone,
+                    'mode': 1,
+                    'Submit': 'Valider'
+                }
+            )
+        except:
+            raise ZoneError
+
+class _BookMyNameClient(object):
+    """
+    Encapsulates all communication with BookMyName.
+    """
+
+    def __init__(self, login, password, session):
+        self.manager = Manager(login, password, session)
+        self.domains_and_ids = {}
+
+    def login(self):
+        """
+        Log in and pre-fetch registered domains and ids
+
+        This method can be called several times; only the first one will
+        cause an actual login and domain+id list fetch.
+        """
+        if len(self.domains_and_ids) == 0:
+            try:
+                self.manager.log_in()
+                try:
+                    self.domains_and_ids = self.manager.get_domains_and_ids()
+                except Error as e:
+                    raise errors.PluginError(
+                        'Error getting BookMyName domains: {0}'.format(e)
+                    )
+            except Error as e:
+                raise errors.PluginError(
+                    'Error logging into BookMyName: {0}'.format(e)
+                )
+
+    def get_registered_domain(self, domain):
+        """
+        Return the registered domain matching the provided domain if any
+
+        A registered domain matches the provided domain if the provided
+        domain ends with the registered domain.
+
+        This method will log in if needed.
+        """
+        self.login()
+        for registered_domain in self.domains_and_ids:
```

Comment:
How about something like:

```python
domains = [registered for registered in self.domains_and_ids if domain.endswith(registered)]
if len(domains) == 0:
    raise errors.PluginError(
    // ... no registered domain
    )

if len(domains) > 1:
    raise errors.PluginError(
    // ... too many registered domains ...
    )

return domains[0]
```

I can't think of a case where there _would be_ multiple matching domains, but it seems better to handle that explicitly than just pick the first we encounter.

Reply:
Ok

Reply:
Done